### PR TITLE
Add some sparse operators

### PR DESCRIPTION
**Pull request by chaihua483**

### CODE REVIEW

**Comment by chaihua483** [comment]
test

**Comment by drpngx** [comment]
@benoitsteiner WDYT?

**Comment by chaihua483** [comment]
With this `op`, it might be easy to implement the gradient of `SparseReduceSumSparse`.

```
import tensorflow.contrib.sparse_ops.python.ops.sparse_ops as spops
@ops.RegisterGradient("SparseReduceSumSparse")
def _SparseReduceSumSparseGrad(op, *grads):
    output_tensor = sparse_tensor.SparseTensor(op.outputs[0], grads[1], op.outputs[2])
    input_tensor = sparse_tensor.SparseTensor(op.inputs[0], op.inputs[1], op.inputs[2])

    grad = spops.sparse_tile_like(output_tensor, input_tensor, op.inputs[3]).values

    return (None, grad,
            None, None)
```

I was wondering if it is proper adding this gradient directly into the `contrib` package?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Can you add all this code to tensorflow.contrib.layers instead of its own new package?

also:
* update tf contrib cmake build to ensure it gets built on windows
* remove all commented out code
* run clang format to clean up the code.

**Comment by chaihua483** [comment]
@ebrevdo Thanks for your reply! I'd love to make changes recently.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by chaihua483** [comment]
@ebrevdo I did just as your suggestions. Please remind me if there is anything else to work on. Thank you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
Please resolve conflicts then we can negotiate like merge. Thanks!

**Review by ebrevdo — DISMISSED** [review]

File: tensorflow/contrib/layers/python/ops/sparse_tile_like_op.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from tensorflow.python.platform import resource_loader
+from tensorflow.python.framework.sparse_tensor import SparseTensor
+
+_sparse_ops = loader.load_op_library(
+    resource_loader.get_path_to_datafile("_sparse_tile_like_op.so"))
+
+
+def sparse_tile_like(sp_a, sp_b, axes):
+  """
+  This `Op` is used for tile a `1D matrix` to `2D matrix` or `2D matrix`
```
Comment (drpngx) [inline tensorflow/contrib/layers/python/ops/sparse_tile_like_op.py]: You'll need to fix the docstring to pass the linter. You need something like
```
  """One-liner

  More verbose description.

  Args:
    sp_a: what it is, e.g. tensor of what type and shape if applicable
    sp_b: ditto
    axes: ditto

  Returns:
     What is returned.
"""
```

File: tensorflow/contrib/layers/ops/sparse_tile_like_op.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    index = [0, 3]
+            [4, 1]
+            [4, 3]
+            [6, 0]
+            [6, 3]
+    values = [1, 2, 2, 3, 3]
+    shape = [10, 10]
+```
+
+As you might see, the `op` will forget all the
```
Comment (drpngx) [inline tensorflow/contrib/layers/ops/sparse_tile_like_op.cc]: please finish the writing

File: tensorflow/contrib/layers/ops/sparse_tile_like_op.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    .Input("b_input_indices: int64")
+    .Input("b_input_values: T")
+    .Input("b_input_shape: int64")
+    .Input("axes: int32")
+    .Output("output_indices: int64")
+    .Output("output_values: T")
+    .Output("output_shape: int64")
+    .Attr("T: realnumbertype")
+    .SetShapeFn(shape_inference::UnknownShape)
+    .Doc(R"doc(
```
Comment (drpngx) [inline tensorflow/contrib/layers/ops/sparse_tile_like_op.cc]: You now need to put a file in the `api_def` directory to document and generate the python.

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    OP_REQUIRES_OK(ctx, ctx->input("a_input_shape", &a_shape_t));
+    OP_REQUIRES_OK(ctx, ctx->input("b_input_indices", &b_indices_t));
+    OP_REQUIRES_OK(ctx, ctx->input("b_input_values", &b_values_t));
+    OP_REQUIRES_OK(ctx, ctx->input("b_input_shape", &b_shape_t));
+    OP_REQUIRES_OK(ctx, ctx->input("axes", &axes_t));
+
+    ValidateInput(ctx, a_indices_t, a_values_t, a_shape_t, b_indices_t,
+                  b_values_t, b_shape_t, axes_t);
+
+    // set values for out_values
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: Here and everywhere: capitalize and punctuate comments.

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    OP_REQUIRES(
+        context,
+        b_shape_t->vec<int64>().size() == a_shape_t->vec<int64>().size() + 1,
+        errors::InvalidArgument(
+            "shape of tensor a should be one dim less than tensor b, but got ",
+            a_shape_t->shape().DebugString(), " and ",
+            b_shape_t->shape().DebugString()));
+  }
+
+};  // class
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: remove comment

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  // (1) Calculate the shape after reduction.
+  auto sp_shape = sp.shape();
+  std::vector<int64> out_dim_sizes;
+  out_dim_sizes = sp.PickDims(reduction.group_by_dims);
+
+  reduction.reduced_shape = TensorShape(out_dim_sizes);
+  return reduction;
+}
+
+std::vector<int64> getRowOfTensor(const Tensor t, const int64 i) {
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: Use `GetRowOfTensor`

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@chaihua483 could you pull rebase and push again?

**Comment by chaihua483** [comment]
Sorry for the delay.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+#include "tensorflow/core/framework/tensor_util.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/lib/strings/str_util.h"
+#include "tensorflow/core/util/sparse/sparse_tensor.h"
+
+using tensorflow::sparse::SparseTensor;
+
+namespace tensorflow {
+
+struct ReduceDetails {
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: Enclose in `namespace {}`

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  // The dimensions to call Reorder() with.
+  std::vector<int64> reorder_dims;
+  // The dimensions to call group() with after Reorder().
+  std::vector<int64> group_by_dims;
+  // The shape after reduction.
+  TensorShape reduced_shape;
+};
+
+ReduceDetails SparseTensorReduceHelper(const SparseTensor& sp,
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: Do not return a struct, use `axes_slice, ReduceDetails* reduction`.

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+  // Now append the rest of the axes (the complement of group_by_dims_);
+  // result is used by Reorder().
+  reduction.reorder_dims = reduction.group_by_dims;
+  std::set_difference(perm.begin(), perm.end(), reduction.group_by_dims.begin(),
+                      reduction.group_by_dims.end(),
+                      std::back_inserter(reduction.reorder_dims));
+
+  // (1) Calculate the shape after reduction.
+  auto sp_shape = sp.shape();
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: unused

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+  std::vector<int64> out_dim_sizes;
+  out_dim_sizes = sp.PickDims(reduction.group_by_dims);
+
+  reduction.reduced_shape = TensorShape(out_dim_sizes);
+  return reduction;
+}
+
+std::vector<int64> GetRowOfTensor(const Tensor t, const int64 i) {
+  std::vector<int64> res;
+  auto t_values = t.matrix<int64>();
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: const auto

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  // (1) Calculate the shape after reduction.
+  auto sp_shape = sp.shape();
+  std::vector<int64> out_dim_sizes;
+  out_dim_sizes = sp.PickDims(reduction.group_by_dims);
+
+  reduction.reduced_shape = TensorShape(out_dim_sizes);
+  return reduction;
+}
+
+std::vector<int64> GetRowOfTensor(const Tensor t, const int64 i) {
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: please don't return `std::vector`s.

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    OP_REQUIRES_OK(ctx, ctx->input("b_input_shape", &b_shape_t));
+    OP_REQUIRES_OK(ctx, ctx->input("axes", &axes_t));
+
+    ValidateInput(ctx, a_indices_t, a_values_t, a_shape_t, b_indices_t,
+                  b_values_t, b_shape_t, axes_t);
+
+    // Set values for out_values.
+    std::vector<int64> group_axes;
+    std::vector<int64> reduction_a;
+    auto axes = axes_t->flat<int32>();
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: const

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    int64 h = 0;
+    int64 output_shape0 = 0;
+    int64 tmp_count = 0;
+    std::vector<int64> output_ids;
+    auto a_values = a_values_t->vec<T>();
+
+    for (const auto& g : sp_b.group(group_axes)) {
+      std::vector<int64> group = g.group();
+      std::vector<int64> row = GetRowOfTensor(sp_a.indices(), h);
+      int64 g_indice_len = g.values<T>().dimension(0);
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: const

File: tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+      while (row < group) {
```
Comment (drpngx) [inline tensorflow/contrib/layers/kernels/sparse_tile_like_kernel.cc]: huh? Does that build?

**Review by drpngx — CHANGES_REQUESTED** [review]
Some cosmetic comments. I stopped at some point, not sure if it's building?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by chaihua483** [comment]
#14964 It seems to have been resolved.

**Pull request closed without merging**